profiling/src/profiling.rs

```diff
@@ -550,85 +566,127 @@ impl Uploader {
-                        info!("Successfully uploaded profile (HTTP {}).", status)
+        loop {
+            /* Since profiling uploads are going over the Internet and not just
+             * the local network, it would be ideal if they were the lowest
+             * priority message, but crossbeam selects at random.
+             * todo: fix fork message priority.
+             */
+            select! {
+                recv(self.fork_receiver) -> message => match message {
+                    Ok(_) => {
+                        // First, wait for every thread to finish what they are currently doing.
+                        self.fork_barrier.wait();
+                        // Then, wait for the fork to be completed.
+                        self.fork_barrier.wait();
                     }
-                }
-                Err(err) => {
-                    warn!("Failed to upload profile: {}", err)
-                }
-            });
+                    _ => {
+                        trace!("Fork channel closed; joining upload thread.");
+                        break;
+                    }
+                },
+
+                recv(self.upload_receiver) -> message => match message {
+                    Ok(upload_message) => match Self::upload(upload_message) {
+                        Ok(status) => {
+                            if status >= 400 {
+                                warn!(
+                                    "Unexpected HTTP status when sending profile (HTTP {}).",
+                                    status
+                                )
+                            } else {
+                                info!("Successfully uploaded profile (HTTP {}).", status)
+                            }
+                        }
+                        Err(err) => {
+                            warn!("Failed to upload profile: {}", err)
+                        }
+                    },
+                    _ => {
+                        trace!("No more upload messages to handle; joining thread.");
+                        break;
+                    }
+
+                },
+            }
+        }
     }
 }
 
 impl Profiler {
     pub fn new() -> Self {
+        let fork_barrier = Arc::new(Barrier::new(3));
+        let (fork_sender0, fork_receiver0) = crossbeam_channel::bounded(1);
+        let vm_interrupts = Arc::new(Mutex::new(Vec::with_capacity(1)));
         let (message_sender, message_receiver) = crossbeam_channel::bounded(100);
-        let (time_collector_cancel_sender, time_collector_cancel_receiver) =
-            crossbeam_channel::bounded(0);
-        let vm_interrupt_lock = Arc::new(Mutex::new(HashSet::with_capacity(1)));
         let (upload_sender, upload_receiver) = crossbeam_channel::bounded(UPLOAD_CHANNEL_CAPACITY);
+        let (fork_sender1, fork_receiver1) = crossbeam_channel::bounded(1);
         let time_collector = TimeCollector {
+            fork_barrier: fork_barrier.clone(),
```

Comment:
This was slightly confusing (for a non rust user), but they will basically share the same barrier when using clone (if I'm reading this right).

Reply:
The clone is for the `Arc<_>`, it's not cloning the barrier itself (which wouldn't make sense and Rust probably prevents). Since 3 threads share ownership of the barrier object, it needs to be refcounted (or proven to outlive all the threads, which is hard).